components/profile/heartbeats.rs

```diff
@@ -2,61 +2,42 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-
 use heartbeats_simple::HeartbeatPow as Heartbeat;
-use heartbeats_simple::HeartbeatPowContext as HeartbeatContext;
 use profile_traits::time::ProfilerCategory;
+use self::synchronized_heartbeat::{heartbeat_window_callback, lock_and_work};
 use servo_config::opts;
 use std::collections::HashMap;
 use std::env::var_os;
 use std::error::Error;
 use std::fs::File;
 use std::path::Path;
-use std::sync::atomic::{ATOMIC_BOOL_INIT, AtomicBool, Ordering};
-
-
-static mut HBS: Option<*mut HashMap<ProfilerCategory, Heartbeat>> = None;
-
-// unfortunately can't encompass the actual hashmap in a Mutex (Heartbeat isn't Send/Sync), so we'll use a spinlock
-static mut HBS_SPINLOCK: AtomicBool = ATOMIC_BOOL_INIT;
-
-fn lock_and_work<F, R>(work: F) -> R where F: FnOnce() -> R {
-    unsafe {
-        while HBS_SPINLOCK.compare_and_swap(false, true, Ordering::SeqCst) {}
-    }
-    let result = work();
-    unsafe {
-        HBS_SPINLOCK.store(false, Ordering::SeqCst);
-    }
-    result
-}
 
 /// Initialize heartbeats
 pub fn init() {
-    let mut hbs: Box<HashMap<ProfilerCategory, Heartbeat>> = Box::new(HashMap::new());
-    maybe_create_heartbeat(&mut hbs, ProfilerCategory::ApplicationHeartbeat);
-    lock_and_work(||
-        unsafe {
-            HBS = Some(Box::into_raw(hbs));
+    lock_and_work(|hbs_opt|
+        if (*hbs_opt).is_none() {
+            let mut hbs: Box<HashMap<ProfilerCategory, Heartbeat>> = Box::new(HashMap::new());
+            maybe_create_heartbeat(&mut hbs, ProfilerCategory::ApplicationHeartbeat);
+            *hbs_opt = Some(Box::into_raw(hbs))
         }
     );
 }
 
 /// Log regmaining buffer data and cleanup heartbeats
 pub fn cleanup() {
-    let hbs_opt: Option<Box<HashMap<ProfilerCategory, Heartbeat>>> = lock_and_work(||
-        unsafe {
-            match HBS {
-                Some(hbs_ptr) => {
-                    let ret = Some(Box::from_raw(hbs_ptr));
-                    HBS = None;
-                    ret
-                },
-                None => None,
-            }
+    let hbs_opt_box: Option<Box<HashMap<ProfilerCategory, Heartbeat>>> = lock_and_work(|hbs_opt|
+        match *hbs_opt {
```

Comment:
I see what you're getting at, though that produced an error:
```
30 |             hbs_opt.take().map(Box::from_raw)
   |                            ^^^ the trait `std::ops::FnOnce<(*mut std::collections::HashMap<profile_traits::time::ProfilerCategory, heartbeats_simple::HeartbeatPow>,)>` is not implemented for `unsafe fn(*mut _) -> std::boxed::Box<_> {<std::boxed::Box<T>><_>::from_raw}`
```
It works if I give it a closure though:
```rust
    let hbs_opt_box: Option<Box<HashMap<ProfilerCategory, Heartbeat>>> = lock_and_work(|hbs_opt|
        hbs_opt.take().map(|hbs_ptr| unsafe {Box::from_raw(hbs_ptr)})
    );
```

The `take` documentation seems to support this approach, but it's not obvious to the the layman (me), so I'll double check - you're certain this will set the `hbs_opt` (and thus `HBS`) option to None afterward?